Static assertion failure due to mismatch of execution policy and functor execution space

I have the following routine, which implements `sort_pairs()` using `kokkos::BinSort`:
```
template < typename EXEC, typename KEY_TYPE, typename VAL_TYPE >
inline void sort_pairs( KEY_TYPE* keys, VAL_TYPE* vals, index_t N )
{
  using kokkos_execspace = typename execution_space< EXEC >::kokkos_execspace;
  using kokkos_memspace  = typename execution_space< EXEC >::kokkos_memspace;
  using RangePolicy = typename execution_space< EXEC >::loop_policy;

  using KeyViewType = Kokkos::View< KEY_TYPE*, kokkos_memspace >;
  using ValViewType = Kokkos::View< VAL_TYPE*, kokkos_memspace >;
  KeyViewType keys_view( keys, N );
  ValViewType vals_view( vals, N );

  using minmax_t = Kokkos::MinMaxScalar<typename KeyViewType::non_const_value_type>;
  minmax_t result;
  Kokkos::MinMax<typename KeyViewType::non_const_value_type> reducer(result);
  Kokkos::parallel_reduce(
    "MinMaxReduce"
  , RangePolicy(0,N)
  , KOKKOS_LAMBDA(const index_t& idx, minmax_t& minmax)
    {
      const KEY_TYPE key = keys[ idx ];
      if ( key < minmax.min_val) minmax.min_val = key;
      if ( key > minmax.max_val) minmax.max_val = key;
    }
  , reducer );

  const index_t max_bins = static_cast< index_t >(0.5 * N);
  using CompType = Kokkos::BinOp1D<KeyViewType>;
  CompType comparator( max_bins, result.min_val, result.max_val );
  Kokkos::BinSort<KeyViewType, CompType> Sorter( keys_view, comparator, true );
  Sorter.create_permute_vector( kokkos_execspace() );
  Sorter.sort( kokkos_execspace(), keys_view );
  Sorter.sort( kokkos_execspace(), vals_view );

  kokkos_execspace().fence();
}
```

The code works with `Kokkos::Serial`, however, when I enable `Kokkos::OpenMP`, I get the following static assert failure:
```
kokkos/core/src/impl/Kokkos_FunctorAnalysis.hpp:119:71: error: static assertion failed: Execution Policy and Functor execution space must match
  119 |   static_assert(!policy_has_space::value || !functor_has_space::value ||
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
  120 |                     std::is_same<typename policy_has_space::type,
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~      
  121 |                                  typename functor_has_space::type>::value,
      |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The stopgap solution provided by @crtrott on a corresponding discussion on Slack is to define `KeyViewType` as follows:
```
using KeyViewType = Kokkos::View< KEY_TYPE*, Kokkos::Device<kokkos_execspace, kokkos_memspace> >;
```

However, it would be nice if that can be handled inside Kokkos instead, if possible.

--- a/algorithms/src/Kokkos_Sort.hpp
+++ b/algorithms/src/Kokkos_Sort.hpp
@@ -220,6 +220,11 @@ class BinSort {
         range_begin(range_begin_),
         range_end(range_end_),
         sort_within_bins(sort_within_bins_) {
+    static_assert(
+        Kokkos::SpaceAccessibility<ExecutionSpace,
+                                   typename Space::memory_space>::accessible,
+        "The provided execution space must be able to access the memory space "
+        "BinSort was initialized with!");
     if (bin_op.max_bins() <= 0)
       Kokkos::abort(
           "The number of bins in the BinSortOp object must be greater than 0!");
@@ -255,6 +260,12 @@ class BinSort {
   // array. Can be called again if keys changed
   template <class ExecutionSpace = exec_space>
   void create_permute_vector(const ExecutionSpace& exec = exec_space{}) {
+    static_assert(
+        Kokkos::SpaceAccessibility<ExecutionSpace,
+                                   typename Space::memory_space>::accessible,
+        "The provided execution space must be able to access the memory space "
+        "BinSort was initialized with!");
+
     const size_t len = range_end - range_begin;
     Kokkos::parallel_for(
         "Kokkos::Sort::BinCount",
@@ -284,6 +295,17 @@ class BinSort {
   template <class ExecutionSpace, class ValuesViewType>
   void sort(const ExecutionSpace& exec, ValuesViewType const& values,
             int values_range_begin, int values_range_end) const {
+    static_assert(
+        Kokkos::SpaceAccessibility<ExecutionSpace,
+                                   typename Space::memory_space>::accessible,
+        "The provided execution space must be able to access the memory space "
+        "BinSort was initialized with!");
+    static_assert(
+        Kokkos::SpaceAccessibility<
+            ExecutionSpace, typename ValuesViewType::memory_space>::accessible,
+        "The provided execution space must be able to access the memory space "
+        "of the View argument!");
+
     using scratch_view_type =
         Kokkos::View<typename ValuesViewType::data_type,
                      typename ValuesViewType::array_layout,

--- a/algorithms/src/Kokkos_Sort.hpp
+++ b/algorithms/src/Kokkos_Sort.hpp
@@ -151,8 +151,8 @@ class BinSort {
     }
   };
 
-  using execution_space = typename Space::execution_space;
-  using bin_op_type     = BinSortOp;
+  using exec_space  = typename Space::execution_space;
+  using bin_op_type = BinSortOp;
 
   struct bin_count_tag {};
   struct bin_offset_tag {};
@@ -238,7 +238,7 @@ class BinSort {
 
   BinSort(const_key_view_type keys_, int range_begin_, int range_end_,
           BinSortOp bin_op_, bool sort_within_bins_ = false)
-      : BinSort(execution_space{}, keys_, range_begin_, range_end_, bin_op_,
+      : BinSort(exec_space{}, keys_, range_begin_, range_end_, bin_op_,
                 sort_within_bins_) {}
 
   template <typename ExecutionSpace>
@@ -248,13 +248,13 @@ class BinSort {
 
   BinSort(const_key_view_type keys_, BinSortOp bin_op_,
           bool sort_within_bins_ = false)
-      : BinSort(execution_space{}, keys_, bin_op_, sort_within_bins_) {}
+      : BinSort(exec_space{}, keys_, bin_op_, sort_within_bins_) {}
 
   //----------------------------------------
   // Create the permutation vector, the bin_offset array and the bin_count
   // array. Can be called again if keys changed
-  template <class ExecutionSpace = execution_space>
-  void create_permute_vector(const ExecutionSpace& exec = execution_space{}) {
+  template <class ExecutionSpace = exec_space>
+  void create_permute_vector(const ExecutionSpace& exec = exec_space{}) {
     const size_t len = range_end - range_begin;
     Kokkos::parallel_for(
         "Kokkos::Sort::BinCount",
@@ -343,7 +343,7 @@ class BinSort {
   template <class ValuesViewType>
   void sort(ValuesViewType const& values, int values_range_begin,
             int values_range_end) const {
-    execution_space exec;
+    exec_space exec;
     sort(exec, values, values_range_begin, values_range_end);
     exec.fence("Kokkos::Sort: fence after sorting");
   }

--- a/algorithms/src/Kokkos_Sort.hpp
+++ b/algorithms/src/Kokkos_Sort.hpp
@@ -151,6 +151,9 @@ class BinSort {
     }
   };
 
+  // Naming this alias "execution_space" would be problematic since it would be
+  // considered as execution space for the various functors which might use
+  // another execution space through sort() or create_permute_vector().
   using exec_space  = typename Space::execution_space;
   using bin_op_type = BinSortOp;
 

